### PR TITLE
Cadc 8514 - PUT with transaction

### PR DESCRIPTION
**Pull request by andamian**
Implements the client side of the mechanism described in https://github.com/pdowler/storage-inventory/blob/master/minoc/PutTransaction.md

Implemented in `BaseWsClient` class so that it can later be ported to `vos`

### CODE REVIEW

File: cadcutils/cadcutils/net/ws.py [hunk 1]
```diff
@@ -351,6 +364,99 @@ def head(self, resource=None, **kwargs):
...
+                    commit_headers = {HTTP_TRANS: transaction_id,
+                                      HTTP_LENGTH: '0'}
+                    response = self._get_session().put(
+                        url,
+                        headers=commit_headers,
+                        verify=self.verify)
+                    if response.status_code != requests.codes.created:  # 201
+                        # might need to try to roll it back at this point
+                        # but unsure how common this case is
+                        raise IOError(
```
Comment (SharonGoliath) [inline cadcutils/cadcutils/net/ws.py:428]: Does this statement also have to cover the case where the transfer failed for some (possibly transient) reason, or is a code other than 201 always a programming BUG, and should result in an OSError being raised?

File: cadcutils/cadcutils/net/ws.py [hunk 2]
```diff
@@ -351,6 +364,99 @@ def head(self, resource=None, **kwargs):
...
+                            'BUG: Could not commit PUT transaction {} for {}'.
+                            format(transaction_id, url))
+                    else:
+                        return
+                else:
+                    self.logger.debug(
+                        'Mismatched md5 checksum on PUT for URL {}. Rollback '
+                        'and re-try'.format(url))
```
Comment (SharonGoliath) [inline cadcutils/cadcutils/net/ws.py:436]: The 'rollback' part of this logging message is done here. Should the logging message say something like 'User must re-try.' (no retry apparent to me in the code).

File: cadcutils/cadcutils/net/ws.py [hunk 3]
```diff
@@ -351,6 +364,99 @@ def head(self, resource=None, **kwargs):
...
+                    rollback_headers = {HTTP_TRANS: transaction_id}
+                    response = self._get_session().delete(
+                        url,
+                        headers=rollback_headers,
+                        verify=self.verify)
+                    if response.status_code == requests.codes.no_content:
+                        raise exceptions.TransferException(
+                            'MD5 checksum mismatch. Transaction rolled back')
+                    else:
+                        raise IOError('BUG: Could not rollback PUT transaction'
```
Comment (SharonGoliath) [inline cadcutils/cadcutils/net/ws.py:446]: Transient error handling or always a BUG?

**Review by SharonGoliath — APPROVED** [review]

File: cadcutils/cadcutils/net/ws.py (hunk 1 above)
Comment (andamian) [reply cadcutils/cadcutils/net/ws.py:428]: Transient errors are dealt with in the `RetrySession` which re-try the `PUT` on transient errors before returning.

File: cadcutils/cadcutils/net/ws.py [hunk 4]
```diff
@@ -351,6 +364,99 @@ def head(self, resource=None, **kwargs):
...
+                                      ' {} for {}'.format(transaction_id, url))
+            else:
+                raise IOError(
+                    'BUG: Expected HTTP 202 with PUT with transactions. '
```
[the remark this answers is not in the document]
Comment (andamian) [reply cadcutils/cadcutils/net/ws.py:450]: Transient errors are dealt with in the RetrySession which re-try the PUT on transient errors before returning. At this point, only 201 is accepted.

**Pull request merged**